### PR TITLE
riscv: use g_running_task store current regs

### PR DESCRIPTION
**Pull request by hujun260**
## Summary
This commit fixes the regression from https://github.com/apache/nuttx/pull/13561

In order to determine whether a context switch has occurred, we can use g_running_task to store the current regs. This allows us to compare the current register state with the previously stored state to identify if a context switch has taken place.

## Impact
riscv

## Testing
rv-virt:knsh
rv-virt:knsh64
rv-virt:smp

### CODE REVIEW

**Comment by nuttxpr** [comment]
[**\[Experimental Bot, please feedback here\]**](https://github.com/search?q=repo%3Aapache%2Fnuttx+13552&type=issues)



The PR summary is missing some information, and the Testing section is incomplete. 

Here's a breakdown:

**Summary - Needs Improvement**

* **What functional part of the code is being changed?**  Be specific. Mention the files, functions, or modules affected. 
* **How does the change exactly work (what will change and how)?**  Provide a technical explanation of your solution.  How does using `g_running_task` to store registers help determine a context switch?

**Impact - Needs Improvement**

* **Is new feature added? Is existing feature changed?** - This seems like a bug fix, so clearly state that.
* **Impact on user (will user need to adapt to change)?** - Even if the answer is NO, you should still briefly explain why.
*  The other "Impact" sections may not all be relevant. It's okay to remove sections if there's genuinely no impact.

**Testing - Incomplete**

* **Build Host(s):** Specify your build host details.
* **Testing logs before change / Testing logs after change:** You need to provide actual logs demonstrating the problem before your fix and the improvement afterward.

**To improve this PR:**

1. **Expand the Summary:**  Clearly describe what part of the code is affected and the technical details of your fix.
2. **Complete the Impact section:**  Address all points, even if it's to say "NO" and provide a brief justification. 
3. **Provide Actual Testing Logs:**  Show the issue before your change and the successful result after.

By providing this additional information, you'll make it easier for reviewers to understand and approve your PR.

**Review by lupyuen — APPROVED** [review]
Tested OK with OSTest on [rv-virt:knsh](https://gist.github.com/lupyuen/09707162e1c628393418bc35c1696b2c), [rv-virt:knsh64](https://gist.github.com/lupyuen/6cb6b3b710e8e804de790a3ad0fc8072) and [milkv_duos:nsh](https://gist.github.com/lupyuen/d48ceaf12d88946639374cf5e0adf574). Thanks!

**Comment by lupyuen** [comment]
Hi @tmedicci would you need to test this on ESP32 RISC-V? Thanks!

**Comment by lupyuen** [comment]
@hujun260 I rebooted and retested OSTest, 10 times on rv-virt:knsh64. Out of the 10 tests, 1 test failed at vfork. Would you know what's causing this? Thanks! https://gist.github.com/lupyuen/cebee9e69f23047476a7a5305f595fa6
```text
user_main: vfork() test
[   71.520000] _assert: Current Version: NuttX  0.0.0 7b5ca0ec41 Sep 24 2024 12:41:49 risc-v
[   71.520000] _assert: Assertion failed : at file: common/riscv_fork.c:133 task: ostest process: ostest 0xc000001a
[   71.520000] up_dump_register: EPC: 00000000802133e4
[   71.520000] up_dump_register: A0: 00000000804072b0 A1: 0000000000000085 A2: 0000000200042020 A3: 0000000080408560
[   71.520000] up_dump_register: A4: 000000000000000a A5: 0000000000000004 A6: 0000000000000000 A7: 0000000000000000
[   71.520000] up_dump_register: T0: 0000000080206898 T1: 0000000000000007 T2: 0000000000000000 T3: 00000000c0206070
[   71.520000] up_dump_register: T4: 00000000c0206068 T5: 00000000c000b8f9 T6: 000000000000003d
[   71.520000] up_dump_register: S0: 0000000000000000 S1: 0000000080409180 S2: 0000000080409180 S3: 0000000000000000
[   71.520000] up_dump_register: S4: 0000000200042020 S5: 0000000000000000 S6: 0000000080219328 S7: 0000000000000000
[   71.520000] up_dump_register: S8: 0000000000000006 S9: 00000000804074d0 S10: 0000000000000085 S11: 0000000080407550
[   71.520000] up_dump_register: SP: 000000008040c850 FP: 0000000000000000 TP: 0000000080409180 RA: 00000000802133e4
[   71.520000] dump_stack: Kernel Stack:
[   71.520000] dump_stack:   base: 0x8040c030
[   71.520000] dump_stack:   size: 00003072
[   71.520000] dump_stack:     sp: 0x8040c850
[   71.520000] stack_dump: 0x8040c810: 0000000000000009 00000000804092d8 000000008021be18 0000000000000000 00000000c0202070 0000000080409180 0000000000000000 0000000080213666
[   71.520000] stack_dump: 0x8040c850: 00000000c000001a fffffffffffffffc 0000000000006010 00000000c0200000 0000000080409180 00000000804072b0 0000000000000000 0000000080219328
[   71.520000] stack_dump: 0x8040c890: 0000000000000085 ffff00587474754e 0000000000000c10 0000000080408180 0000000080408180 0000000080206af0 00000000c0206000 302e300080408180
[   71.520000] stack_dump: 0x8040c8d0: 000000008000302e 00000000802086ca 3061633562370000 7065532031346365 3432303220343220 343a31343a323120 0000000080400039 0000000080203352
[   71.520000] stack_dump: 0x8040c910: 7369720000000000 0000000000762d63 0000000000000000 000000008040cc40 000000008040cf40 0000000000000000 0000000000000000 0000000000000000
[   71.520000] stack_dump: 0x8040c950: 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000005 0000000200042022 0000000080409180 00000000c000919a
[   71.520000] stack_dump: 0x8040c990: ffffffffffffffda 0000000080206912 ffffffffffffffda 000000008020a32c ffffffffffffffff 0000000000000000 0000000000000005 0000000000000020
[   71.520000] stack_dump: 0x8040c9d0: 0000000000000008 00000000c000919a ffffffffffffffda 00000000802068a0 0000000000000008 0000000080200940 0000000000000000 0000000080200972
[   71.520000] stack_dump: 0x8040ca10: 0000000200042020 00000000802001a2 00000000c000919a 00000000c0008af8 00000000c0203ef0 0000000080409180 0000000000000000 0000000000000000
[   71.520000] stack_dump: 0x8040ca50: 00000000c01017b0 0000000000000000 00000000c0202040 0000000000000000 0000000000000022 00000000c02003d8 00000000c0200448 0000000000000018
[   71.520000] stack_dump: 0x8040ca90: 000000000000000a 00000000c0101f4e 0000000000000009 0000000000000000 0000000000000005 00000000c0010a48 0000000000000005 0000000000000000
[   71.520000] stack_dump: 0x8040cad0: 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000 00000000c0008146 00000000c000b6f8
[   71.520000] stack_dump: 0x8040cb10: 00000000c000b8f9 000000000000003d 0000000200042020 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000
[   71.520000] stack_dump: 0x8040cb50: 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000
[   71.520000] stack_dump: 0x8040cb90: 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000
[   71.520000] stack_dump: 0x8040cbd0: 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000 00000000802034b2
[   71.520000] stack_dump: 0x8040cc10: 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000
ostest_main: Exiting with status 256
```

**Comment by hujun260** [comment]
> @hujun260 I rebooted and retested OSTest, 10 times on rv-virt:knsh64. Out of the 10 tests, 1 test failed at vfork. Would you know what's causing this? Thanks! https://gist.github.com/lupyuen/cebee9e69f23047476a7a5305f595fa6
> 
> ```
> user_main: vfork() test
> [   71.520000] _assert: Current Version: NuttX  0.0.0 7b5ca0ec41 Sep 24 2024 12:41:49 risc-v
> [   71.520000] _assert: Assertion failed : at file: common/riscv_fork.c:133 task: ostest process: ostest 0xc000001a
> [   71.520000] up_dump_register: EPC: 00000000802133e4
> [   71.520000] up_dump_register: A0: 00000000804072b0 A1: 0000000000000085 A2: 0000000200042020 A3: 0000000080408560
> [   71.520000] up_dump_register: A4: 000000000000000a A5: 0000000000000004 A6: 0000000000000000 A7: 0000000000000000
> [   71.520000] up_dump_register: T0: 0000000080206898 T1: 0000000000000007 T2: 0000000000000000 T3: 00000000c0206070
> [   71.520000] up_dump_register: T4: 00000000c0206068 T5: 00000000c000b8f9 T6: 000000000000003d
> [   71.520000] up_dump_register: S0: 0000000000000000 S1: 0000000080409180 S2: 0000000080409180 S3: 0000000000000000
> [   71.520000] up_dump_register: S4: 0000000200042020 S5: 0000000000000000 S6: 0000000080219328 S7: 0000000000000000
> [   71.520000] up_dump_register: S8: 0000000000000006 S9: 00000000804074d0 S10: 0000000000000085 S11: 0000000080407550
> [   71.520000] up_dump_register: SP: 000000008040c850 FP: 0000000000000000 TP: 0000000080409180 RA: 00000000802133e4
> [   71.520000] dump_stack: Kernel Stack:
> [   71.520000] dump_stack:   base: 0x8040c030
> [   71.520000] dump_stack:   size: 00003072
> [   71.520000] dump_stack:     sp: 0x8040c850
> [   71.520000] stack_dump: 0x8040c810: 0000000000000009 00000000804092d8 000000008021be18 0000000000000000 00000000c0202070 0000000080409180 0000000000000000 0000000080213666
> [   71.520000] stack_dump: 0x8040c850: 00000000c000001a fffffffffffffffc 0000000000006010 00000000c0200000 0000000080409180 00000000804072b0 0000000000000000 0000000080219328
> [   71.520000] stack_dump: 0x8040c890: 0000000000000085 ffff00587474754e 0000000000000c10 0000000080408180 0000000080408180 0000000080206af0 00000000c0206000 302e300080408180
> [   71.520000] stack_dump: 0x8040c8d0: 000000008000302e 00000000802086ca 3061633562370000 7065532031346365 3432303220343220 343a31343a323120 0000000080400039 0000000080203352
> [   71.520000] stack_dump: 0x8040c910: 7369720000000000 0000000000762d63 0000000000000000 000000008040cc40 000000008040cf40 0000000000000000 0000000000000000 0000000000000000
> [   71.520000] stack_dump: 0x8040c950: 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000005 0000000200042022 0000000080409180 00000000c000919a
> [   71.520000] stack_dump: 0x8040c990: ffffffffffffffda 0000000080206912 ffffffffffffffda 000000008020a32c ffffffffffffffff 0000000000000000 0000000000000005 0000000000000020
> [   71.520000] stack_dump: 0x8040c9d0: 0000000000000008 00000000c000919a ffffffffffffffda 00000000802068a0 0000000000000008 0000000080200940 0000000000000000 0000000080200972
> [   71.520000] stack_dump: 0x8040ca10: 0000000200042020 00000000802001a2 00000000c000919a 00000000c0008af8 00000000c0203ef0 0000000080409180 0000000000000000 0000000000000000
> [   71.520000] stack_dump: 0x8040ca50: 00000000c01017b0 0000000000000000 00000000c0202040 0000000000000000 0000000000000022 00000000c02003d8 00000000c0200448 0000000000000018
> [   71.520000] stack_dump: 0x8040ca90: 000000000000000a 00000000c0101f4e 0000000000000009 0000000000000000 0000000000000005 00000000c0010a48 0000000000000005 0000000000000000
> [   71.520000] stack_dump: 0x8040cad0: 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000 00000000c0008146 00000000c000b6f8
> [   71.520000] stack_dump: 0x8040cb10: 00000000c000b8f9 000000000000003d 0000000200042020 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000
> [   71.520000] stack_dump: 0x8040cb50: 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000
> [   71.520000] stack_dump: 0x8040cb90: 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000
> [   71.520000] stack_dump: 0x8040cbd0: 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000 00000000802034b2
> [   71.520000] stack_dump: 0x8040cc10: 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000
> ostest_main: Exiting with status 256
> ```

ok i'll try to reproduce

**Comment by lupyuen** [comment]
@hujun260 I did some more Stress Testing:
1. For the code in this PR: Out of 20 reboots and retests, 2 failed with the vfork error. So the failure rate is 10%.
2. For the code dated 12 Sep, 2 weeks ago [(9bee10f)](https://github.com/apache/nuttx/tree/9bee10f05e4516902395266480f551a2f94c1326): Out of 200 reboots and retests, all 200 succeeded without vfork error.

So this vfork problem might be introduced in the past 2 weeks?

**Review by lupyuen — CHANGES_REQUESTED** [review]
Let's do more analysis on the vfork issue, to ensure it's not caused by this PR. Thanks!

**Comment by hujun260** [comment]
> Let's do more analysis on the vfork issue, to ensure it's not caused by this PR. Thanks!
ok

**Comment by hujun260** [comment]
> [lupyuen](/lupyuen)

@lupyuen  Could you send me your stress testing script? I can occasionally reproduce the issue on my side, but the efficiency is relatively low.

**Comment by lupyuen** [comment]
@hujun260 Here's my Stress Test Script that will run 20 times: https://github.com/lupyuen/nuttx-riscv64/blob/main/special-qemu-riscv-knsh64.sh

Please change https://github.com/apache/nuttx/tree/master to https://github.com/hujun260/nuttx/tree/apache_11

File: arch/risc-v/src/common/riscv_doirq.c [hunk 1]
```diff
@@ -70,10 +71,14 @@ uintreg_t *riscv_doirq(int irq, uintreg_t *regs)
   if (irq >= RISCV_IRQ_ECALLU && irq <= RISCV_IRQ_ECALLM)
     {
       regs[REG_EPC] += 4;
+      if (regs[REG_A0] != SYS_restore_context)
+        {
+          (*running_task)->xcp.regs = regs;
```
Comment (xiaoxiang781216) [inline arch/risc-v/src/common/riscv_doirq.c:76]: why not always set xcp.regs to regs?

File: arch/risc-v/src/common/supervisor/riscv_perform_syscall.c [hunk 2]
```diff
@@ -37,19 +37,21 @@
...
   up_set_current_regs(regs);
 
   /* Run the system call handler (swint) */
 
-  ret = riscv_swint(0, regs, NULL);
+  riscv_swint(0, regs, NULL);
+  tcb = this_task();
 
-  if (ret == SWINT_CONTEXT_SWITCH)
+  if ((*running_task)->xcp.regs != tcb->xcp.regs)
```
Comment (xiaoxiang781216) [inline arch/risc-v/src/common/supervisor/riscv_perform_syscall.c:54]: why not compare *running_task with tcb directly

File: arch/risc-v/src/common/riscv_swint.c [hunk 3]
```diff
@@ -199,7 +199,7 @@ uintptr_t dispatch_syscall(unsigned int nbr, uintptr_t parm1,
 int riscv_swint(int irq, void *context, void *arg)
 {
   uintreg_t *regs = (uintreg_t *)context;
-  uintreg_t *new_regs = regs;
```
Comment (xiaoxiang781216) [inline arch/risc-v/src/common/riscv_swint.c:202]: why need change

**Review by lupyuen — APPROVED** [review]
Tested OK with OSTest on [rv-virt:knsh64](https://gist.github.com/lupyuen/1359330caa433c84009c3563dcded8bb) and [milkv_duos:nsh](https://gist.github.com/lupyuen/8173f10e9543d5f2593f5e52d1c08b40).

[Stressed Tested](https://github.com/lupyuen/nuttx-riscv64/blob/main/special-qemu-riscv-knsh64.sh#L20-L30) 200 times (reboot + ostest) on rv-virt:knsh64. All 200 times succeeded 🎉 Thanks :-)

File: arch/risc-v/src/common/riscv_fork.c [hunk 4]
```diff
@@ -147,15 +149,15 @@ pid_t riscv_fork(const struct fork_s *context)
   /* Set up frame for context and copy the parent's user context there */
 
   memcpy((void *)(newsp - XCPTCONTEXT_SIZE),
```
Comment (pussuw) [inline arch/risc-v/src/common/riscv_fork.c:151]: The error is here, newsp is using the child's _user stack_ to store the context. It should use the kernel stack.

Also, the fork() error is not related to this PR whatsoever, please make it into a separate patch and a separate PR.

Comment (pussuw) [reply arch/risc-v/src/common/riscv_fork.c:151]: SMP mode fork is also broken due to #12812. When the child runs, the context is read from the user stack but the wrong mappings are in use -> the child crashes due to page fault.

Please, let me fix the fork() issue @hujun260 @lupyuen

File: arch/risc-v/src/common/riscv_fork.c [hunk 5]
```diff
@@ -102,6 +103,7 @@
 
 pid_t riscv_fork(const struct fork_s *context)
 {
+  uintreg_t *parent_regs = (uintreg_t *)riscv_percpu_get_regs();
```
Comment (pussuw) [inline arch/risc-v/src/common/riscv_fork.c:106]: Please remove the fork() related fixes from this patch, as they are not related to the topic of this PR or the patch.

Comment (hujun260) [reply arch/risc-v/src/common/riscv_fork.c]: ok

**Review by lupyuen — CHANGES_REQUESTED** [review]
Thanks @hujun260! Let's wait for https://github.com/apache/nuttx/pull/13633 to be merged. Then we can rebase and retest your PR. Thank you so much :-)

**Comment by lupyuen** [comment]
Hi @hujun260: https://github.com/apache/nuttx/pull/13633 has been merged. Could you please rebase to the master branch and retest your PR? Thanks!

**Comment by hujun260** [comment]
> Thanks @hujun260! Let's wait for #13633 to be merged. Then we can rebase and retest your PR. Thank you so much :-)

ok

**Review by lupyuen — APPROVED** [review]
[Stress Test of rv-virt:knsh64](https://github.com/lupyuen/nuttx-riscv64/blob/main/special-qemu-riscv-knsh64.sh#L20-L30) (reboot qemu + restart ostest) succeeded for all 200 iterations. Also tested OK with OSTest on [milkv_duos:nsh](https://gist.github.com/lupyuen/7f4c54df0c8ebc6f2645fb921d87b98e). Thank you so much!

**Review by xiaoxiang781216 — APPROVED** [review]

File: arch/arm/src/armv8-r/Toolchain.defs [hunk 6]
```diff
@@ -23,10 +23,10 @@ ARCHCPUFLAGS += -march=armv8-r
 ifeq ($(CONFIG_ARCH_FPU),y)
   LLVM_ABITYPE := eabihf
 
-  ifeq ($(CONFIG_ARCH_CORTEXR52)$(CONFIG_ARM_NEON),yy)
```
Comment (pussuw) [inline arch/arm/src/armv8-r/Toolchain.defs:26]: The commit message and these changes are now in conflict with the PR topic. Modifications to armv8 and risc-v but the topic is for risc-v only.

Please restore the original patch and remove the armv8 stuff

Comment (hujun260) [reply arch/arm/src/armv8-r/Toolchain.defs]: ok

Comment (pussuw) [reply arch/arm/src/armv8-r/Toolchain.defs]: Thanks, LGTM now

**Review by lupyuen — APPROVED** [review]
Retested: [Stress Test of rv-virt:knsh64](https://github.com/lupyuen/nuttx-riscv64/blob/main/special-qemu-riscv-knsh64.sh#L20-L30) (reboot qemu + restart ostest) succeeded for all 200 iterations. Also tested OK with OSTest on [milkv_duos:nsh](https://gist.github.com/lupyuen/42e907472ef07cc4177ad4c0e23a6065). Thank you so much!

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by lupyuen** [comment]
FYI All the Daily Tests are successful now 🎉
- [milkv_duos:nsh](https://github.com/lupyuen/nuttx-sg2000/releases/tag/nuttx-sg2000-2024-09-28)
- [rv-virt:nsh64](https://github.com/lupyuen/nuttx-riscv64/releases/tag/qemu-riscv-nsh64-2024-09-28)
- [rv-virt:knsh64](https://github.com/lupyuen/nuttx-riscv64/releases/tag/qemu-riscv-knsh64-2024-09-28)
